### PR TITLE
Feat: arrow csv decimal256

### LINKED ISSUE #3474: arrow-csv: support decimal256
I noticed that

https://github.com/apache/arrow-rs/blob/master/arrow-csv/src/reader/mod.rs

only seems to support decimal128 (ended up there via an error message from csv2arrow).
Is there a reason for that? If not, would be amazing if someone could implement support.
Decimal256 is pretty important since its the only way to represent bigints (well big-ish ints) in arrow afaict.

### PR DESCRIPTION
**Pull request by suxiaogang223**
# Which issue does this PR close?

Closes #3474 

# Rationale for this change
 

arrow-csv only seems to support decimal128

# What changes are included in this PR?

now the arrow-csv can support decimal256

# Are there any user-facing changes?

nothing

### CODE REVIEW

**Review by tustvold — APPROVED** [review]
This looks good to me, thank you.

File: arrow-csv/src/reader/mod.rs [hunk 1]
```diff
@@ -814,6 +818,41 @@ fn build_decimal_array(
     ))
 }
 
+//TODO: is possible to use generic function replace this?
```
[the remark this answers is not in the document]
Comment (LiShiZhensPi) [reply arrow-csv/src/reader/mod.rs:821]: Thanks for your help, I'll fix these soon🌝

**Comment by LiShiZhensPi** [comment]
To support operations `ArrowNativeTypeOp` in generic functions, I change the code to `type Native: ArrowNativeType + ArrowNativeTypeOp` in privimitive_array.rs. I'm not sure if this is correct. Pay attention to this when reviewing the code🤓

File: arrow-array/src/array/primitive_array.rs [hunk 2]
```diff
@@ -233,7 +233,8 @@ pub type Decimal256Array = PrimitiveArray<Decimal256Type>;
 /// static-typed nature of rust types ([`ArrowNativeType`]) for all types that implement [`ArrowNativeType`].
 pub trait ArrowPrimitiveType: 'static {
     /// Corresponding Rust native type for the primitive type.
-    type Native: ArrowNativeType;
+    // FIXME: this may affect other code, is this correct?
+    type Native: ArrowNativeType + ArrowNativeTypeOp;
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:237]:
```suggestion
    type Native: ArrowNativeTypeOp;
```

`ArrowNativeType` is implied by `ArrowNativeTypeOp`.

I think this change is fine, but tagging @viirya who added `ArrowNativeTypeOp` originally

**Review by tustvold — APPROVED** [review]
We should probably return an error instead of panicking on overflow, but this looks good to me :+1:

File: arrow-array/src/array/primitive_array.rs (hunk 2 above)
Comment (LiShiZhensPi) [reply arrow-array/src/array/primitive_array.rs]: yes, It is indeed better to write this way

**Comment by LiShiZhensPi** [comment]
now the function return error instead of panic🤓
 https://github.com/apache/arrow-rs/pull/3711#pullrequestreview-1297811625

File: arrow-csv/src/reader/mod.rs [hunk 3]
```diff
@@ -781,22 +795,22 @@ fn parse_bool(string: &str) -> Option<bool> {
 }
 
 // parse the column string to an Arrow Array
-fn build_decimal_array(
+fn build_decimal_array<T: DecimalType>(
```
Comment (viirya) [inline arrow-csv/src/reader/mod.rs:798]: 👍

File: arrow-csv/src/reader/mod.rs [hunk 4]
```diff
@@ -1318,7 +1339,7 @@ mod tests {
         let schema = Schema::new(vec![
             Field::new("city", DataType::Utf8, false),
             Field::new("lat", DataType::Decimal128(38, 6), false),
-            Field::new("lng", DataType::Decimal128(38, 6), false),
+            Field::new("lng", DataType::Decimal256(38, 6), false),
```
Comment (viirya) [inline arrow-csv/src/reader/mod.rs:1342]: Maybe use a larger precision/scale for `Decimal256`.

File: arrow-csv/src/reader/mod.rs [hunk 5]
```diff
@@ -1788,20 +1826,20 @@ mod tests {
...
+            let result = parse_decimal_with_parameter::<Decimal128Type>(s, 20, 3);
             assert_eq!(
                 format!("Parser error: can't parse the string value {s} to decimal"),
                 result.unwrap_err().to_string()
             );
         }
         let overflow_parse_tests = ["12345678", "12345678.9", "99999999.99"];
         for s in overflow_parse_tests {
-            let result = parse_decimal_with_parameter(s, 10, 3);
+            let result = parse_decimal_with_parameter::<Decimal128Type>(s, 10, 3);
```
Comment (viirya) [inline arrow-csv/src/reader/mod.rs:1842]: Perhaps adding test for `Decimal256Type`.

File: arrow-csv/src/writer.rs [hunk 6]
```diff
@@ -406,6 +408,59 @@ sed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000,23:46:03,foo
         assert_eq!(expected.to_string(), String::from_utf8(buffer).unwrap());
     }
 
+    #[test]
+    fn test_write_csv_decimal() {
+        let schema = Schema::new(vec![
+            Field::new("c1", DataType::Decimal128(38, 6), true),
+            Field::new("c2", DataType::Decimal256(38, 6), true),
```
Comment (viirya) [inline arrow-csv/src/writer.rs:415]: ditto. Maybe using larger precision/scale for Decimal256.

**Review by viirya — APPROVED** [review]

File: arrow-csv/src/reader/mod.rs (hunk 4 above)
Comment (LiShiZhensPi) [reply arrow-csv/src/reader/mod.rs:1342]: yes, I found the `DECIMAL256_MAX_PRECISION` is 76, so maybe change it to `Decimal256(76, 6)` is ok?

Comment (viirya) [reply arrow-csv/src/reader/mod.rs]: Sounds good to me.

**Comment by viirya** [comment]
Thank you. This looks good.

**Comment by tustvold** [comment]
Thank you again

**Pull request merged**